### PR TITLE
List append

### PR DESCRIPTION
**Pull request by weilewei**
@hkaiser adding `append` function for the list. Could you please review it and suggest if I miss something? Thank you so much. 
Wei

### CODE REVIEW

File: src/plugins/listops/append_operation.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        hpx::util::make_tuple("append",
+            std::vector<std::string>{"append(_1, _2)"},
+            &create_append_operation, &create_primitive<append_operation>)};
+
+    ///////////////////////////////////////////////////////////////////////////
+    void append_operation::append_element(
+        std::vector<primitive_argument_type> & result,
+        primitive_argument_type && rhs) const
+    {
+        if (is_list_operand_strict(rhs))
```
Comment (hkaiser) [inline src/plugins/listops/append_operation.cpp]: What are the intended semantics here? Which of the following is what you'd like to achieve:
```
    append(list(1, 2, 3), list(4, 5, 6)) == list(1, 2, 3, 4, 5, 6)
```
or
```
    append(list(1, 2, 3), list(4, 5, 6)) == list(1, 2, 3, list(4, 5, 6))
```
?

File: tests/unit/plugins/listops/append_operation.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+///////////////////////////////////////////////////////////////////////////////
+phylanx::execution_tree::compiler::function compile(std::string const& code)
+{
+    phylanx::execution_tree::compiler::function_list snippets;
+    phylanx::execution_tree::compiler::environment env =
+        phylanx::execution_tree::compiler::default_environment();
+
+    return phylanx::execution_tree::compile(code, snippets, env);
+}
```
Comment (hkaiser) [inline tests/unit/plugins/listops/append_operation.cpp]: Please verify this code against current master as the compiler API has changed slightly. You will have to adapt this test accordingly.

File: tests/unit/plugins/listops/append_operation.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+///////////////////////////////////////////////////////////////////////////////
+void test_append_operation(std::string const& code,
+    std::string const& expected_str)
+{
+    HPX_TEST_EQ(compile(code)(), compile(expected_str)());
+}
```
Comment (hkaiser) [inline tests/unit/plugins/listops/append_operation.cpp]: The same applies here.

File: tests/unit/plugins/listops/append_operation.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+///////////////////////////////////////////////////////////////////////////////
+int main(int argc, char* argv[])
+{
+    test_append_operation("append( make_list(1), 2 )", "list(1, 2)");
+    test_append_operation("append( make_list(), make_list() )", "list()");
+    test_append_operation(
+        "append( make_list(1, 2), make_list() )", "list(1, 2)");
+
```
Comment (hkaiser) [inline tests/unit/plugins/listops/append_operation.cpp]: Please add more test cases beyond the corner cases already in place.

File: src/plugins/listops/append_operation.cpp (hunk 1 above)
Comment (hkaiser) [reply src/plugins/listops/append_operation.cpp]: FWIW, Python exposes the latter semantics:
```
>>> l = [1, 2, 3]       
>>> l.append([3, 4, 5]) 
>>> l                   
[1, 2, 3, [3, 4, 5]]    
```

**Comment by hkaiser** [comment]
Very nice! Thanks a lot!

File: tests/unit/plugins/listops/append_operation.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    test_append_operation("append( make_list(1, 42), make_list(3, 4) )",
+        "list(1, 42, list(3, 4))");
+    test_append_operation("append( make_list(), make_list() )", "list(list())");
+    test_append_operation(
+        "append( make_list(1, 42), make_list() )", "list(1, 42, list())");
+    test_append_operation(
+        "append( make_list(), make_list(1, 42) )", "list(list(1, 42))");
+    test_append_operation("append( make_list(1, 2), 42 )", "list(1, 2, 42)");
+    test_append_operation("append( make_list(1, 2), hstack(3, 42) )",
+        "list(1, 2, hstack([3, 42]))");
```
Comment (hkaiser) [inline tests/unit/plugins/listops/append_operation.cpp]: This doesn't look right. Is it really resulting in `hstack([3, 42])`? Shouldn't it evaluate to `hstack(3, 42)`?

**Comment by weilewei** [comment]
@hkaiser I updated my code this afternoon and now all the checks passed. I will later create a ticket for the `hstack` thing we talked about today. Could you please check if this PR looks good to you? Thanks.

**Comment by hkaiser** [comment]
Could you please rebase your changes on top of master, please? This PR currently has way too many unrelated changes to review properly.
